packaging/os/yum.py

```diff
@@ -646,60 +698,40 @@ def install(module, items, repoq, yum_basecmd, conf_file, en_repos, dis_repos):
+            # determine if this is a downgrade, and then use the yum downgrade command to install the RPM.
+
+            # Get the NEVRA of the requested package
+            (name, ver, rel, epoch, arch) = splitFilename(pkglist.pop())
+
+            # Check if any version of the requested package is installed
+            inst_pkgs = is_installed(module, repoq, name, conf_file, en_repos=en_repos, dis_repos=dis_repos, is_pkg=True)
+            if inst_pkgs:
+                (cur_name, cur_ver, cur_rel, cur_epoch, cur_arch) = splitFilename(inst_pkgs[0])
+                compare = compareEVR((cur_epoch, cur_ver, cur_rel), (epoch, ver, rel))
+                if compare > 0:
+                    downgrade = True
+
             # if not - then pass in the spec as what to install
             # we could get here if nothing provides it but that's not
             # the error we're catching here
             pkg = spec
 
-        pkgs.append(pkg)
+        if downgrade:
+            downgrade_pkgs.append(pkg)
+        else:
+            pkgs.append(pkg)
 
+    if downgrade_pkgs:
+        res = exec_install(module, items, 'downgrade', downgrade_pkgs, res, yum_basecmd, tempdir)
     if pkgs:
-        cmd = yum_basecmd + ['install'] + pkgs
-
-        if module.check_mode:
-            # Remove rpms downloaded for EL5 via url
-            try:
-                shutil.rmtree(tempdir)
-            except Exception, e:
-                module.fail_json(msg="Failure deleting temp directory %s, %s" % (tempdir, e))
-
-            module.exit_json(changed=True, results=res['results'], changes=dict(installed=pkgs))
-
-        changed = True
-
-        lang_env = dict(LANG='C', LC_ALL='C', LC_MESSAGES='C')
-        rc, out, err = module.run_command(cmd, environ_update=lang_env)
-
-        if (rc == 1):
-            for spec in items:
-                # Fail on invalid urls:
-                if ('://' in spec and ('No package %s available.' % spec in out or 'Cannot open: %s. Skipping.' % spec in err)):
-                    err = 'Package at %s could not be installed' % spec
-                    module.fail_json(changed=False,msg=err,rc=1)
-        if (rc != 0 and 'Nothing to do' in err) or 'Nothing to do' in out:
-            # avoid failing in the 'Nothing To Do' case
-            # this may happen with an URL spec.
-            # for an already installed group,
-            # we get rc = 0 and 'Nothing to do' in out, not in err.
-            rc = 0
-            err = ''
-            out = '%s: Nothing to do' % spec
-            changed = False
-
-        res['rc'] = rc
-        res['results'].append(out)
-        res['msg'] += err
-
-        # FIXME - if we did an install - go and check the rpmdb to see if it actually installed
-        # look for each pkg in rpmdb
-        # look for each pkg via obsoletes
-
-        # Record change
-        res['changed'] = changed
+        res = exec_install(module, items, 'install', pkgs, res, yum_basecmd, tempdir)
 
```

Comment:
As far as I can tell, there's no way to include the downgrades and upgrades in the same yum transaction. Maybe test if the end state matches the spec and error out of it doesn't? That still doesn't really solve the problem.